src/main/java/org/elasticsearch/gateway/DanglingIndicesState.java

```diff
@@ -0,0 +1,177 @@
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.env.NodeEnvironment;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * The dangling indices state is responsible for finding new dangling indices (indices that have
+ * their state written on disk, but don't exists in the metadata of the cluster), and importing
+ * them into the cluster.
+ */
+public class DanglingIndicesState extends AbstractComponent {
+
+    private final NodeEnvironment nodeEnv;
+    private final MetaStatePersistence metaStatePersistence;
+    private final LocalAllocateDangledIndices allocateDangledIndices;
+
+    private final Set<String> danglingIndices = ConcurrentCollections.newConcurrentSet();
+
+    @Inject
+    public DanglingIndicesState(Settings settings, NodeEnvironment nodeEnv, MetaStatePersistence metaStatePersistence,
+                                LocalAllocateDangledIndices allocateDangledIndices) {
+        super(settings);
+        this.nodeEnv = nodeEnv;
+        this.metaStatePersistence = metaStatePersistence;
+        this.allocateDangledIndices = allocateDangledIndices;
+    }
+
+    /**
+     * Process dangling indices based on the provided meta data, handling cleanup, finding
+     * new dangling indices, and allocating outstanding ones.
+     */
+    public void processDanglingIndices(MetaData metaData) {
+        if (nodeEnv.hasNodeFile() == false) {
+            return;
+        }
+        cleanupAllocatedDangledIndices(metaData);
+        findNewDanglingIndices(metaData);
+        allocateDanglingIndices(outstandingDanglingIndicesToProcess());
+    }
+
+    /**
+     * The current set of dangling indices.
+     */
+    Set<String> getDanglingIndices() {
+        return ImmutableSet.copyOf(danglingIndices);
+    }
+
+    /**
+     * Cleans dangling indices if they are already allocated on the provided meta data.
+     */
+    void cleanupAllocatedDangledIndices(MetaData metaData) {
+        for (String danglingIndex : danglingIndices) {
+            if (metaData.hasIndex(danglingIndex)) {
+                logger.debug("[{}] no longer dangling (created), removing from dangling list", danglingIndex);
+                danglingIndices.remove(danglingIndex);
+            }
+        }
+    }
+
+    /**
+     * Finds new dangling indices by iterating over the indices and trying to find indices
+     * that have state on disk, but are not part of the provided meta data, or not detected
+     * as dangled already.
+     */
+    void findNewDanglingIndices(MetaData metaData) {
+        final Set<String> indices;
+        try {
+            indices = nodeEnv.findAllIndices();
+        } catch (Throwable e) {
+            logger.warn("failed to list dangling indices", e);
+            return;
+        }
+
+        for (String indexName : indices) {
+            if (metaData.hasIndex(indexName)) { // if we have the index on the metadata, its not dangled
```

Comment:
can we remove the continue here and just do a `if((metaData.hasIndex(indexName) || danglingIndices.contains(indexName)) == false)` ?